[For review and improvement] E-mail requirement and Password change features

D00MK1D opened the pull request:
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

Implemented features in the account route for mandatory email registration and password change, allowing users to control their own passwords.

Currently, the route is only accessible by the user through my webpage project, which sends emails and calls the API to use "mail_token" for changes.

#Future implementations
- Change email route (Currently made on webpage by direct db UPDATE which is not so safe)
- SMTP functionality for this project (With Mailkit)
- Email address validation for new accounts


# Checklist:
- [ X ] The project compiles
- [ X ] The PR targets `develop` branch

D00MK1D committed:
Reset Password Route

D00MK1D committed:
Initial e-mail change/verification

D00MK1D committed:
nullifies tokens on account creation

D00MK1D committed:
Rename GenerateLoginToken() to GenerateToken() since it'll be multipurpose by now

D00MK1D committed:
improved password change

Token for this operation can expire by time forcing a new token to be generated and token is cleaned from DB when used

D00MK1D committed:
first MailToken to future "verify e-mail" operation

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
Insert blank values on mail_token and password_token to new players instead of DbNull

D00MK1D committed:
Merge branch 'develop' of https://github.com/D00MK1D/Arrowgene.DragonsDogmaOnline into develop

D00MK1D committed:
Improve password reset and token validation logic

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
better email validation regex

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
Merge branch 'develop' of https://github.com/D00MK1D/Arrowgene.DragonsDogmaOnline into develop

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
email checking and cog page working now

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

alborrajo commented on Arrowgene.Ddon.Database/Sql/Core/DdonSqlDbAccount.cs:
```diff
@@ -30,9 +32,9 @@ public partial class DdonSqlDb : SqlDb
-        account.MailToken = Convert.ToBase64String(Encoding.UTF8.GetBytes(mail));
-        account.PasswordToken = Convert.ToBase64String(Encoding.UTF8.GetBytes(name));
-        account.LoginToken = Convert.ToBase64String(Encoding.UTF8.GetBytes(name));
+        account.MailToken = GameToken.GenerateToken();
```
Why is this one not being set to null like the rest?

D00MK1D replied:
Will be used as a token to validate e-mail of new accounts

alborrajo replied:
I still don't think it should be the job of a database class to generate tokens. I think it'd be better if this was passed as an argument to the function

alborrajo commented on Arrowgene.Ddon.Database/Sql/Core/DdonSqlDbAccount.cs:
```diff
@@ -77,6 +79,19 @@ public partial class DdonSqlDb : SqlDb
 
+    public override Account? SelectAccountByEmail(string email)
+    {
+        email = email.ToLowerInvariant();
```
What if the data in the column does have uppercase characters

D00MK1D replied:
I can test this scenario but isn't email already inserted in lower case?
Upper case would occur if it's manually inserted data and select it with lower case would return this data normally.
Or i'm i wrong?

alborrajo replied:
You could have this be part of the query. Like `WHERE UPPER(mail) = UPPER(@mail)`

D00MK1D replied:
Like this you say?
`private static readonly string SqlSelectAccountByEmail = $"SELECT \"id\", {BuildQueryField(AccountFields)} FROM \"account\" WHERE UPPER(\"mail\")=UPPER(@mail);";`

alborrajo replied:
Yes, but please test first if such a query would work, i'm mentioning it by memory

alborrajo commented on Arrowgene.Ddon.Database/Sql/Core/DdonSqlDbAccount.cs:
```diff
@@ -109,8 +124,8 @@ public override bool UpdateAccount(Account account)
-            AddParameter(command, "@mail_token", account.MailToken);
-            AddParameter(command, "@password_token", account.PasswordToken);
+            AddParameter(command, "@mail_token", string.IsNullOrWhiteSpace(account.MailToken) ? "" : account.MailToken);
+            AddParameter(command, "@password_token", string.IsNullOrWhiteSpace(account.PasswordToken) ? "" : account.PasswordToken);
```
Why not make these fields nullable in the database?

D00MK1D replied:
For some unknown fcking reason i was having 0x0000000c when these ones were dbnull.
This was a workaround that i forgot to deal properly before draft this PR

alborrajo commented on Arrowgene.Ddon.GameServer/Handler/ConnectionLoginHandler.cs:
```diff
@@ -86,7 +86,7 @@ public override S2CConnectionLoginRes Handle(GameClient client, C2SConnectionLog
 
             // update login token for client
-            // client.Account.LoginToken = GameToken.GenerateLoginToken();
+            // client.Account.LoginToken = GameToken.GenerateToken();
```
Dead code?

D00MK1D replied:
It was already there so i think yes, got renamed by CTRL + H

alborrajo replied:
Better delete these lines then

D00MK1D replied:
Deleted

alborrajo commented on Arrowgene.Ddon.Shared/Model/GameToken.cs:
```diff
@@ -10,7 +10,7 @@ public class GameToken
         public const int GameTokenLength = 20;
 
-        public static string GenerateLoginToken()
+        public static string GenerateToken()
```
Why the rename?

D00MK1D replied:
https://discord.com/channels/596777418178691073/596777784928632858/1381682935794761798

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -118,15 +145,39 @@ public override async Task<WebResponse> Post(WebRequest request)
                     res.Message = "Account created";
                     break;
+
+                case "reset":
```
I'm not too sure I understand the flow. If there's already a password token, all this will do nothing and it'll go directly to the ResetPassword call? I don't think the intent is very clear

D00MK1D replied:
This same route will be called 2 times:

 - First it'll generate the token that must be sent to user 
 - Second when user send this token to route with the new password it'll actually change the password
 
 Finally it's erased from DB

D00MK1D replied:
Resolved splitting to "recover" and "reset"

Recover will generate token and send it via email
Reset will allow a password change with token as validation

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -164,10 +222,50 @@ private string CreateToken(string name, string password)
+        private Account ResetPassword(string name, string password, string passwordToken)
+        {
+            Account account = _database.SelectAccountByName(name);
+            if (account == null && account.PasswordToken != passwordToken && account.MailVerified)
```
This if statement wont do what you intend it to do, and most dangerously, it'll allow changing anyones password even if the token doesnt match

D00MK1D replied:
I think my tests didn't covered this situation, i'll check

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -164,10 +222,50 @@ private string CreateToken(string name, string password)
+            }
+
+            account.PasswordToken = GameToken.GenerateToken();
+            account.MailVerifiedAt = DateTime.UtcNow;
```
I'm not sure this field should be updated

D00MK1D replied:
Why shouldn't?

D00MK1D replied:
Ok... it is in the wrong place

alborrajo commented on Arrowgene.Ddon.WebServer/Files/www/web/index.html:
```diff
@@ -3,76 +3,126 @@
+                        <div><span class="label">Token:</span> ${tokenText}</div>
+                    `;
+
+                    document.getElementById("response_area").innerHTML = html;
```
Generating HTML dynamically like that can be dangerous but considering we won't be showing other user's content i think we can let it slide

D00MK1D replied:
i will rework it entirelly, this index is AI code that i used just to mess around with my routes directly from server.
in fact this was not intended to be in this PR

D00MK1D replied:
Still AI coded but not dynamically generated anymore

alborrajo commented on Arrowgene.Ddon.WebServer/Files/www/web/index.html:
```diff
@@ -3,76 +3,126 @@
+                },
+            };
+
+            DDON.setup("http://localhost:52099/api/");
```
While we're at it i think we can change it to simply /api and maybe that way it will work out of the box for hosts that arent localhost

D00MK1D replied:
It worked as you said

D00MK1D committed:
Email verify route

D00MK1D committed:
Reset password route fix

-Fixed the ability to change any account password with any random token.

-Better readability

D00MK1D committed:
Update DdonSqlDbAccount.cs

reverted changes

D00MK1D committed:
Merge branch 'develop' into develop

D00MK1D committed:
Merge branch 'sebastian-heinz:develop' into develop

D00MK1D committed:
CORS Middleware so you can do api thing on javascript pages

D00MK1D committed:
Shows message in-game if someone forcibly try to log with token not provided by server

D00MK1D committed:
Update Arrowgene.WebServer

D00MK1D committed:
Better ResetPassword

D00MK1D committed:
Little index adjust

D00MK1D committed:
SMTP Functionality

D00MK1D committed:
A standard logo suggestion

D00MK1D committed:
removed test code from AccountRoute

D00MK1D committed:
Changed ServerUrl to DomainUrl for better understanding

D00MK1D committed:
Update IndexRoute.cs

For some reason, every route was resulting in index, even if i "/kjsndiouanufe.aksmowuinergwe/aosdoqasm" the IndexRoute would be showed, so i renamed it

D00MK1D committed:
Some logic corrections on password change

D00MK1D committed:
Passing labels to english (sorry...)

alborrajo commented on Arrowgene.Ddon.LoginServer/Handler/ClientLoginHandler.cs:
```diff
@@ -54,6 +54,8 @@ public override L2CLoginRes Handle(LoginClient client, C2LLoginReq request)
                     if (account == null)
                     {
+                        L2CEjectionNtc message = new L2CEjectionNtc { Message = "Invalid login attempt" };
+                        client.Send(message);
```
Is this needed? 🤔

D00MK1D replied:
Not needed, but did this just to clearly show to players that their
`DDO.exe "addr=dd-on.com port=52100 token=Doomkid DL=http://dd-on.com:52099/win/ LVer=03.04.003.20181115.0 RVer=3040008"`
Isn't a valid way to login instead of showing `Error:SomeError(50M3C0D3)`

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -94,39 +123,103 @@ public override async Task<WebResponse> Post(WebRequest request)
                         break;
                     }
 
+                    if (token == "mail")
```
I don't like this. If it returns a token it should be a proper token. Having special magic values feels wrong. Could it not throw an exception, or simply return null like it does for wrong logins?

D00MK1D replied:
I'll properly change this

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -94,39 +123,103 @@ public override async Task<WebResponse> Post(WebRequest request)
                         break;
                     }
 
                     res.Message = "Account created";
```
The message should be updated so it says that the email should be verified before the player can log in

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -164,10 +264,98 @@ private string CreateToken(string name, string password)
+                return null;
+            }
+
+            if (account.PasswordToken != passwordToken)
```
Would this ever happen? Wouldn't it return null since it can't find the account by the provided password token?

D00MK1D replied:
This will not happen anymore indeed, i was using e-mail as parameter to select account before, since i use passwordToken now, this can be removed for sure

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -94,39 +123,103 @@ public override async Task<WebResponse> Post(WebRequest request)
             }
 
             AccountResponse res = new AccountResponse();
-            AccountVerification accountCheck = new(req.Account, req.Password);
```
We can't remove this. Sure, we can't require the email field for every action, but we shouldn't stop verifying the fields altogether

D00MK1D replied:
Even if the result of this is used only to "create" action?
It's now placed in "create" case

alborrajo replied:
This was used to verify that logins contained both account id and password, we should still verify that, and also verify email in the create actions

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -94,39 +123,103 @@ public override async Task<WebResponse> Post(WebRequest request)
+                    break;
+
+                case "reset":
+                    account = ResetPassword(req.Password, req.PasswordToken);
```
For better security we should also require the account id and verify that both the account id and password token matches when resetting the password

D00MK1D replied:
Ok, i'll rollback this then

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -94,39 +123,103 @@ public override async Task<WebResponse> Post(WebRequest request)
+                    break;
+
+                case "verify":
+                    bool verification = VerifyEmail(req.EmailToken);
```
We should also probably require both the account id and the email token

D00MK1D replied:
Ok, i'll rollback this then too

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -94,39 +123,103 @@ public override async Task<WebResponse> Post(WebRequest request)
+                    break;
+
+                case "resend":
+                    account = ResendEmailVerification(req.Email);
```
We should probably also require the account id

D00MK1D replied:
Ok, i'll rollback this then too too

alborrajo replied:
🤔

alborrajo commented on Arrowgene.Ddon.WebServer/CorsMiddleware.cs:
THANK YOU

alborrajo commented on Arrowgene.Ddon.WebServer/CorsMiddleware.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        WebResponse response;
+
+        if (request.Method == WebRequestMethod.Unknown)
```
Why this special case? afaik CORS requests dont have an unknown method, they use the OPTIONS method

D00MK1D replied:
Sadly i had no OPTION

<img width="254" height="221" alt="image" src="https://github.com/user-attachments/assets/28860832-e1e9-4699-a617-50b4dd026249" />

alborrajo replied:
HTTP is a plain text protocol, can't you compare the method against the string "OPTIONS"?

alborrajo commented on Arrowgene.Ddon.WebServer/Files/mail_templates/mail_verify.html:
```diff
@@ -0,0 +1,129 @@
+                    <table class="container" cellpadding="0" cellspacing="0">
+                        <tr>
+                            <td class="header">
+                                <img src="http://{{DomainUrl}}:52099/launcher/logo.png" alt="Logo">
```
I think both the schema and port should be part of the settings, that way admins can host it behind HTTPS only and through whatever port they want. Granted, this template is customizable, so they could edit it, but if we do it this way its one more thing that works out of the box

D00MK1D replied:
Will be done

alborrajo commented on Arrowgene.Ddon.WebServer/Files/mail_templates/mail_verify.html:
```diff
@@ -0,0 +1,129 @@
+                                <hr style="width:inherit;color: #b5b194;"/>
+
+                                <p>
+                                    © {{Year}} Dragon's Dogma Online Team
```
Absolutely remove this, we dont own the copyright to this game 😭

D00MK1D replied:
"WDT Team" sounds any good?

alborrajo replied:
I don't think we'd need a copyright to begin with

alborrajo commented on Arrowgene.Ddon.WebServer/Files/mail_templates/new_account.html:
```diff
@@ -0,0 +1,129 @@
+                                <hr style="width:inherit;color: #b5b194;"/>
+
+                                <p>
+                                    © {{Year}} Dragon's Dogma Online Team
```
ditto

alborrajo commented on Arrowgene.Ddon.WebServer/Files/mail_templates/reset_password.html:
```diff
@@ -0,0 +1,129 @@
+                                <hr style="width:inherit;color: #b5b194;"/>
+
+                                <p>
+                                    © {{Year}} Dragon's Dogma Online Team
```
ditto

alborrajo commented on Arrowgene.Ddon.WebServer/Files/www/launcher/logo.png:
Remove this too, change it to clip art or DDOn (not the full name) in a fancy font, but dont add this copyrighted image

alborrajo commented on Arrowgene.Ddon.WebServer/Files/www/web/mail_verify.html:
```diff
@@ -0,0 +1,69 @@
+
+    <div id="response_area">
+        <div><span class="label">Mensagem:</span> <span id="resp_message"></span></div>
+        <div><span class="label">Erro:</span> <span id="resp_erro"></span></div>
```
🇧🇷

D00MK1D replied:
Already resolved 🫡

alborrajo commented on Arrowgene.Ddon.WebServer/Files/www/web/reset_password.html:
```diff
@@ -0,0 +1,120 @@
+        </div>
+
+        <div class="field">
+            <button id="button_submit" type="button">Alterar Senha</button>
```
🇧🇷

D00MK1D replied:
Already resolved 🫡

alborrajo commented on Arrowgene.Ddon.WebServer/IndexRoute.cs:
```diff
@@ -1,12 +1,12 @@
     public class IndexRoute : WebRoute
     {
-        public override string Route => "/";
+        public override string Route => "/index";
```
Why is this needed?

D00MK1D replied:
For some reason i couldn't track, every route on 52099 was being catch, even if i typed random shit like 
`http://localhost:52099/kjhadboiu.asldjnoqw9781623896523/adsljdn0qiuwed.skah` for example
It wil AWAYS direct me to "Welcome - Index Page"

I can try to track it down later but that was the shortest way to deal with this and i don't think its a problem since this does nothing

I trust in your judgement

alborrajo replied:
I don't think it's too bad that unknown routes redirect to the index, as long as they dont obscure other routes

D00MK1D replied:
It was obscuring other routes indeed

D00MK1D replied:
Correcting my self, it's not obscuring the routes but this

<img width="1000" height="207" alt="image" src="https://github.com/user-attachments/assets/99cec260-3ab6-4b3a-8364-fe74bd8899d9" />

D00MK1D replied:
Resolved rolling back the Arrowgene.WebServer from 1.1.8 to 1.1.4

alborrajo commented on Arrowgene.Ddon.WebServer/MailSend.cs:
```diff
@@ -0,0 +1,137 @@
+                {
+                    ["{{UserName}}"] = WebUtility.HtmlEncode(account.Name),
+                    ["{{DomainUrl}}"] = baseUrl,
+                    ["{{VerificationLink}}"] = $"http://{baseUrl}:52099/web/verify.html?token={Uri.EscapeDataString(account.MailToken)}",
```
ditto about the schema and port, should probably be configurable

alborrajo commented on Arrowgene.Ddon.WebServer/MailSend.cs:
```diff
@@ -0,0 +1,137 @@
+
+            var baseUrl = _mailSetting.DomainUrl?.TrimEnd('/') ?? throw new InvalidOperationException("ServerUrl is not configured");
+
+            if(mailModel == "new_account" || mailModel == "mail_verify")
```
I don't like that we're allowing any string as a template in the arguments but we're comparing strings to make sure they're only this or that values. Either we change the mailModel to an enum, or we allow it to be anything and dynamically require only what the template is using.

alborrajo commented on Arrowgene.Ddon.WebServer/MailSend.cs:
```diff
@@ -0,0 +1,137 @@
+                {
+                    ["{{UserName}}"] = WebUtility.HtmlEncode(account.Name),
+                    ["{{DomainUrl}}"] = baseUrl,
+                    ["{{PasswordResetLink}}"] = $"http://{baseUrl}:52099/web/reset_password.html?token={Uri.EscapeDataString(account.PasswordToken)}",
```
ditto

alborrajo commented on Arrowgene.Ddon.WebServer/MailSend.cs:
```diff
@@ -0,0 +1,137 @@
+
+        private static string SubjectFor(string mailModel)
+        {
+            return mailModel switch
```
ditto about mailModel being a string vs an enum

alborrajo commented on Arrowgene.Ddon.WebServer/MailSend.cs:
```diff
@@ -0,0 +1,137 @@
+        {
+            return mailModel switch
+            {
+                "new_account" => "Welcome to Dragon's Dogma Online",
```
Dont put the game's name, change to DDOn and/or have the subject be configurable

alborrajo commented on Arrowgene.Ddon.WebServer/MailSend.cs:
```diff
@@ -0,0 +1,137 @@
+                "new_account" => "Welcome to Dragon's Dogma Online",
+                "mail_verify" => "Verify your updated e-mail",
+                "reset_password" => "Password reset",
+                _ => "Dragon's Dogma Online"
```
ditto

D00MK1D committed:
Generate mailToken outside DB class to pass it as an argument

D00MK1D committed:
Removing ConnectionLoginHandler dead code

D00MK1D committed:
Delete logo.png

D00MK1D committed:
Update ClientLoginHandler.cs

D00MK1D committed:
Update DatabaseMigratorTest.cs

alborrajo commented on Arrowgene.Ddon.LoginServer/Handler/ClientLoginHandler.cs:
```diff
@@ -54,6 +54,8 @@ public override L2CLoginRes Handle(LoginClient client, C2LLoginReq request)
                 {
```
This might be over the scope of this PR, but now that we have a launcher everyone uses and it's easier than using the classic start_ddo_local.cmd, it might be time to phase out and remove the option to disable AccountRequired and flows like this

D00MK1D replied:
If you guys agree with that, no problem.

alborrajo reviewed (changes requested):
Now that i think of it, would we want the server to require mail validation by default?  Is there a way to disable it so accounts dont require validation? 🤔

D00MK1D committed:
rollback to WebServer 1.1.4

D00MK1D committed:
Adding account name to previously only token query

D00MK1D committed:
Requested adjusts

D00MK1D committed:
Merge branch 'develop' into develop

D00MK1D committed:
Update CorsMiddleware.cs

D00MK1D committed:
Merge branch 'develop' of https://github.com/D00MK1D/Arrowgene.DragonsDogmaOnline into develop

D00MK1D commented:
Requesting re-review, when possible

D00MK1D committed:
removing unreachable code

alborrajo commented on Arrowgene.Ddon.WebServer/Files/www/web/mail_verify.html:
```diff
@@ -0,0 +1,92 @@
+
+                getTokenFromUrl: function () {
+                    var query = window.location.search.substring(1);
+                    var params = query.split("&");
```
Replace with
```js
const params = new URLSearchParams(window.location.search)
```

alborrajo commented on Arrowgene.Ddon.WebServer/Files/www/web/mail_verify.html:
```diff
@@ -0,0 +1,92 @@
+                    var query = window.location.search.substring(1);
+                    var params = query.split("&");
+
+                    for (var i = 0; i < params.length; i++) {
```
Replace with
```js
return params.get("token")
```

alborrajo commented on Arrowgene.Ddon.WebServer/MailSetting.cs:
```diff
@@ -0,0 +1,111 @@
+            SmtpPassword ??= "p@55VV0rD";
+            FromAddress ??= "no-reply@dd.on";
+            TemplatePath ??= Path.Combine(Util.ExecutingDirectory(), "Files/mail_templates");
+            LogoLink ??= "DDON - Welcome to Ddon!";
```
shouldn't this be an url to an image file, a local path at least

alborrajo commented on Arrowgene.Ddon.WebServer/MailSetting.cs:
```diff
@@ -0,0 +1,111 @@
+            SmtpPassword = "p@55VV0rD";
+            FromAddress = "no-reply@dd.on";
+            TemplatePath = Path.Combine(Util.ExecutingDirectory(), "Files/mail_templates");
+            LogoLink = "DDON - Welcome to Ddon!";
```
ditto

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -77,12 +81,35 @@ public AccountVerification(string username, string password)
+                    return;
+                }
+
+                if (Regex.IsMatch(Email, @"\s"))
```
Isn't this check already covered by the regex below?

D00MK1D committed:
requested changes

D00MK1D committed:
Redundant regex removal

D00MK1D committed:
MailRequired check on "recover" route

D00MK1D committed:
Merge branch 'develop' into develop

alborrajo commented on Arrowgene.Ddon.WebServer/AccountRoute.cs:
```diff
@@ -118,15 +140,102 @@ public override async Task<WebResponse> Post(WebRequest request)
+                    }
+                    catch
+                    {
+                        res.Error = "Verification e-mail could not be sent";
```
Nitpick but i think the message should lead the user to asking for a resend later, otherwise they might not know what to do, only that theyre locked out of the game

D00MK1D replied:
Okay
And i just noted that i didn't did the "resend" page

D00MK1D committed:
Create verify_resend.html

D00MK1D committed:
Update AccountRoute.cs

D00MK1D committed:
Merge branch 'develop' into develop

D00MK1D committed:
Merge branch 'develop' into develop

alborrajo committed:
Merge branch 'develop' into develop
